libdnf/base/goal.cpp

```diff
@@ -596,10 +604,136 @@ void Goal::Impl::add_rpms_to_goal(base::Transaction & transaction) {
+                            rpm::PackageQuery query_id(base, rpm::PackageQuery::ExcludeFlags::IGNORE_EXCLUDES, true);
+                            query_id.add(rpm::Package(base, rpm::PackageId(id)));
+                            query_id.filter_upgrades();
+                            if (query_id.empty()) {
+                                // Report when package with lower version is not installed
+                                std::string name_arch(pool.get_name(id));
+                                name_arch.append(".");
+                                name_arch.append(pool.get_arch(id));
+                                transaction.p_impl->add_resolve_log(
+                                    GoalAction::UPGRADE,
+                                    GoalProblem::ALREADY_INSTALLED,
+                                    settings,
+                                    {pool.get_nevra(id)},
+                                    {name_arch},
+                                    false);
+                            } else {
+                                // Only installed packages with same name, architecture and lower version can be upgraded
+                                ids_upgrades.push_back(id);
+                            }
+                        }
+                    }
+                }
                 rpm_goal.add_upgrade(
                     ids, settings.resolve_best(cfg_main), settings.resolve_clean_requirements_on_remove());
             } break;
+            case GoalAction::DOWNGRADE: {
+                solv::IdQueue ids_downgrades;
+                for (auto id : ids) {
+                    rpm::PackageQuery query(installed);
+                    query.filter_name({pool.get_name(id)});
+                    if (query.empty()) {
+                        // Report when package with the same name is not installed
+                        transaction.p_impl->add_resolve_log(
+                            GoalAction::DOWNGRADE,
+                            GoalProblem::NOT_INSTALLED,
+                            settings,
+                            {pool.get_nevra(id)},
+                            {},
+                            false);
+                    } else {
+                        query.filter_arch({pool.get_arch(id)});
+                        if (query.empty()) {
+                            // Report when package with the same name is installed for a different architecture
+                            transaction.p_impl->add_resolve_log(
+                                GoalAction::DOWNGRADE,
+                                GoalProblem::NOT_INSTALLED_FOR_ARCHITECTURE,
+                                settings,
+                                {pool.get_nevra(id)},
+                                {},
+                                false);
+                        } else {
+                            rpm::PackageQuery query_id(base, rpm::PackageQuery::ExcludeFlags::IGNORE_EXCLUDES, true);
```

Comment:
What about to use `query` like `query.filter_evr(pool.get_evr(id),libdnf::sack::QueryCmp::GT)`

Be honest I really don't understand why we have `pool.get_evr()'. It is not transparent at all. It does not say that ID must be package id and not reldep ID.

Reply:
OK. Changed.